### PR TITLE
Variables

### PR DESCRIPTION
**Pull request by Wolfe-Lyon**
Hello Richard.

I have implemented a new SkoolKit feature! I have enhanced _snapinfo.py_ such that it can now list all variables set in a snapshot.

I started by extracting number and text reading functions into their own modules so that they can be used by both the basic listing mode and the new variable listing mode. I also enhanced the number reading functions to return signed numbers, rather than only unsigned numbers.

There are a number of additional unit tests included in this pull request, covering the changes I have made.

I appreciate that this is quite a large change, so if anything doesn't make sense, or if you would like me to make further modifications, please let me know.

### CODE REVIEW

**Comment by Wolfe-Lyon** [comment]
I've removed the floats from _VariableListerTest_ and replaced them with integers. Floats are already tested in _NumberUtilsTest_.

**Comment by skoolkid** [comment]
Hi Philip

Thanks for adding this feature to snapinfo.py!

Unfortunately you've missed the window for getting it into 5.3, but on the plus side that means we've got plenty of time to get it into shape for 5.4. I've had a quick look over your changes, and I have some comments and suggestions.

* Instead of creating three new modules, I would keep VariableLister and the functions it needs in basic.py (so that everything to do with reading BASIC programs and variables is in one place).
* Likewise, the tests for VariableLister etc. could reside in test_basic.py.
* The snapinfo.py options to list the BASIC program and the variables should not be mutually exclusive - for example, 'snapinfo.py -bv' should list both the BASIC program and its variables; this could be done by incorporating the call to VariableLister inside BasicLister and passing the -b/-v options as boolean flags into the BasicLister constructor (for example).
* Conditions in 'if' statements - indeed expressions anywhere - should not contain redundant brackets.
* Code coverage needs to be 100%!

I'm letting 5.3 soak for a while now while I use it to make updates to the JSW, MM and ROM disassemblies. Once 5.3 is out, we can start looking at merging this request.

**Comment by Wolfe-Lyon** [comment]
Thanks for your comments, Richard.

- I noticed that back in commit ea22f01 that you had moved _BasicLister_ to its own module. It felt natural to me to continue this theme and keep _VariableLister_ in its own module as well. However I'm happy to revise this if you believe it to be appropriate.
- As above.
- I will look into setting this up.
- Redundant brackets will be expunged.
- I noticed the build and test failures as soon as I submitted the pull request. Unfortunately I don't have a CI / testing environment set up so I was not able to catch these issues before submitting the pull request. Hopefully these have all now been addressed.

**Comment by Wolfe-Lyon** [comment]
I have:
- consolidated all basic- and variables-related code into _basic.py_,
- consolidated all basic- and variables-related test code into _test_basic.py_,
- lifted the mutual exclusivity that was on the -b and -v options of _snapinfo.py_ (handling this at the point of argument parsing rather than by introducing dependencies between _VariableLister_ and _BasicLister_),
- removed all redundant outer brackets from 'if' and 'while' statements (but leaving behind functionally redundant brackets in expressions where those brackets enhance readability, in my opinion) and
- ensured that code coverage is 100%

Please let me know of anything else you think needs to be addressed, or of anything I have missed.

**Comment by skoolkid** [comment]
Thanks for making those changes. I'll add comments to the code wherever I think further changes should be made.

File: skoolkit/basic.py [hunk 1]
```diff
@@ -194,24 +254,152 @@ def _get_num_str(self, j):
...
-        self.lspace = True
-        if code < 168 or code == 203 or token[-1] in '#=>':
-            # RND, INKEY$, PI, THEN, '<=', '>=', '<>', 'OPEN #', 'CLOSE #'
-            return token
-        self.lspace = False
-        return token + ' '
+    def list_variables(self, snapshot):
+        lines = ['Variables:']
+        self.snapshot = snapshot
+        i = (snapshot[23627] + 256 * snapshot[23628]) or 23755
```
Comment (skoolkid) [inline skoolkit/basic.py]: Just assume that the variables start at VARS, and let any fallout happen if VARS is not set correctly (and get rid of _skip_basic_line()). I know BasicLister takes PROG to be 23755 if it's set to zero, but that's because I was initially testing it on snapshots created by bin2tap.py, which doesn't set system variables.

Later on I might add options to snapinfo.py for setting PROG and VARS explicitly.

File: skoolkit/basic.py [hunk 2]
```diff
@@ -194,24 +254,152 @@ def _get_num_str(self, j):
...
+        while i < len(snapshot) and snapshot[i] != 128:
+            self.text.lspace = False
+            variable_type = (snapshot[i] & 224)
```
Comment (skoolkid) [inline skoolkit/basic.py]: Redundant brackets.

File: skoolkit/basic.py [hunk 3]
```diff
@@ -194,24 +254,152 @@ def _get_num_str(self, j):
...
+        str_length = get_word(self.snapshot, i + 1)
+        i += 3
+        while str_length > 0:
+            self.text.lspace = True
+            code = self.snapshot[i]
+            line += self.text.get_chars(code)
+            str_length -= 1
+            i += 1
+        line += '"'
+        return i, line
```
Comment (skoolkid) [inline skoolkit/basic.py]: This while loop looks unnecessary. Couldn't you do something like this:

    name = self.text.get_chars((self.snapshot[i] & 31) + 96)
    value = ''.join([self.text.get_chars(c) for c in self.snapshot[i + 3:i + 3 + str_length]])
    line = '(String) {}$="{}"'.format(letter, value)
    return i + 3 + str_length, line

Same goes for the while loops in _get_num_array_var(), _get_long_num_var() and _get_char_array_var().

File: skoolkit/basic.py [hunk 4]
```diff
@@ -194,24 +254,152 @@ def _get_num_str(self, j):
...
+        i += 5
+        line += 'limit={}, '.format(get_number(self.snapshot, i))
+        i += 5
+        line += 'step={}, '.format(get_number(self.snapshot, i))
+        i += 5
+        line += 'line={}, '.format(get_word(self.snapshot, i))
+        i += 2
+        line += 'statement={})'.format(self.snapshot[i])
+        i += 1
+        return i,line
```
Comment (skoolkid) [inline skoolkit/basic.py]: This is unnecessarily verbose. It could be reduced to something like this:

    ...
    line += "{}=".format(self.text.get_chars(letter))
    line += "{} (".format(get_number(self.snapshot, i + 1))
    line += 'limit={}, '.format(get_number(self.snapshot, i + 6))
    ...
    return i + 19, line

Same goes for _get_short_num_var().

File: skoolkit/snapinfo.py [hunk 5]
```diff
@@ -404,6 +405,8 @@ def main(args):
     group = parser.add_argument_group('Options')
     group.add_argument('-b', '--basic', action='store_true',
                        help='List the BASIC program')
+    group.add_argument('-v', '--vars', action='store_true',
+                       help='List variables')
```
Comment (skoolkid) [inline skoolkit/snapinfo.py]: Can you move these two lines below -p/--peek: options should be defined in alphabetical order.

Also use --variables instead of --vars, in case I want to use --vars later for setting VARS!

File: skoolkit/basic.py (hunk 3 above)
Comment (Wolfe-Lyon) [inline skoolkit/basic.py]: I have done this for _get_string_var(), _get_num_array_var() and _get_char_array_var(). However the 'while' loop in _get_long_num_var() is different in that it is not iterating over a fixed range of addresses. Instead it is checking the most significant bit at the current address to decide whether to terminate the string data.

File: skoolkit/basic.py (hunk 2 above)
Comment (Wolfe-Lyon) [inline skoolkit/basic.py]: Removed

File: skoolkit/snapinfo.py (hunk 5 above)
Comment (Wolfe-Lyon) [inline skoolkit/snapinfo.py]: Lines moved below -t/--find-text, assuming you meant alphabetical order with respect to the short argument flags.

File: skoolkit/basic.py (hunk 4 above)
Comment (Wolfe-Lyon) [inline skoolkit/basic.py]: Verbosity reduced.

File: skoolkit/basic.py (hunk 1 above)
Comment (Wolfe-Lyon) [inline skoolkit/basic.py]: Now always uses the value in VARS to determine start address of variables.

I have not yet removed "_skip_basic_line()". I added this to cover the (perhaps unlikely) possibility that one or more lines of basic exist in the variables area. If I remove it, there will be no defined behaviour when a byte of the form 000xxxxx or 001xxxxx is encountered. Is this what we want? Perhaps it would be appropriate to stop processing if a byte of either of these forms is encountered?

**Comment by Wolfe-Lyon** [comment]
Most of the changes you requested have now been made. Please see above for my comments.

File: skoolkit/basic.py [hunk 6]
```diff
@@ -114,14 +115,92 @@
...
+def get_number(snapshot, i):
+    if snapshot[i] == 0:
+        # Small integer (signed)
+        num = _get_integer(snapshot, i)
+    else:
+        # Floating point number (signed)
+        num = _get_float(snapshot, i)
+    return num
+
+def _get_integer(snapshot, i):
```
Comment (skoolkid) [inline skoolkit/basic.py]: This method need not be so long and complicated. Let's assume for now that the sign byte is either 0 (positive) or 255 (negative). Try this:

    if snapshot[i + 1]:
        return get_word(snapshot, i + 2) - 65536
    return get_word(snapshot, i + 2)

File: skoolkit/basic.py [hunk 7]
```diff
@@ -114,14 +115,92 @@
     255: 'COPY'
 }
 
+def get_number(snapshot, i):
+    if snapshot[i] == 0:
```
Comment (skoolkid) [inline skoolkit/basic.py]: There is no need to use the 'num' variable here. Also explicitly testing whether an integer is zero is usually unpythonic.

    if snapshot[i]:
        return _get_float(snapshot, i)
    return _get_integer(snapshot, i)

File: skoolkit/basic.py [hunk 8]
```diff
@@ -114,14 +115,92 @@
...
+    if (snapshot[i + 1] & 128) == 0:
+        sign = 1
+    else:
+        sign = -1
+    mantissa = float(16777216 * (snapshot[i + 1] | 128)
+                     + 65536 * snapshot[i + 2]
+                     + 256 * snapshot[i + 3]
+                     + snapshot[i + 4])
+    num = sign * mantissa * (2 ** exponent)
+    return num
```
Comment (skoolkid) [inline skoolkit/basic.py]: No need to assign the value to 'num' here. Just return the value.

File: skoolkit/basic.py [hunk 9]
```diff
@@ -114,14 +115,92 @@
...
+        self.lspace = False
+        return token + ' '
+
 class BasicLister:
+    def __init__(self):
+        self.text = TextReader()
+
     def list_basic(self, snapshot):
-        lines = []
+        lines = ['Basic listing:']
```
Comment (skoolkid) [inline skoolkit/basic.py]: Leave out this heading. I think the difference between a BASIC listing and the variables is clear enough.

File: skoolkit/basic.py [hunk 10]
```diff
@@ -194,24 +254,112 @@ def _get_num_str(self, j):
...
-        if self.lspace and code >= 197 and token[0] >= 'A':
-            token = ' ' + token
-        self.lspace = True
-        if code < 168 or code == 203 or token[-1] in '#=>':
-            # RND, INKEY$, PI, THEN, '<=', '>=', '<>', 'OPEN #', 'CLOSE #'
-            return token
-        self.lspace = False
-        return token + ' '
+    def list_variables(self, snapshot):
+        lines = ['Variables:']
```
Comment (skoolkid) [inline skoolkit/basic.py]: Leave out this heading.

File: skoolkit/snapinfo.py [hunk 11]
```diff
@@ -426,8 +429,11 @@ def main(args):
         _find_text(infile, namespace.text)
     elif namespace.peek is not None:
         _peek(infile, namespace.peek)
-    elif namespace.basic:
-        print(BasicLister().list_basic(get_snapshot(infile)))
+    elif namespace.basic or namespace.variables:
+        if namespace.basic:
+            print(BasicLister().list_basic(get_snapshot(infile)))
+        if namespace.variables:
+            print(VariableLister().list_variables(get_snapshot(infile)))
```
Comment (skoolkid) [inline skoolkit/snapinfo.py]: It's possible that get_snapshot() will be called twice here - once more than is necessary.

File: skoolkit/basic.py [hunk 12]
```diff
@@ -114,14 +115,92 @@
...
+        num *= -1
+
+    return num
+
+def _get_float(snapshot, i):
+    exponent = snapshot[i] - 160
+    if (snapshot[i + 1] & 128) == 0:
+        sign = 1
+    else:
+        sign = -1
```
Comment (skoolkid) [inline skoolkit/basic.py]: As already mentioned, explicitly testing whether an integer is zero is unpythonic. Also, redundant brackets.

    if snapshot[i + 1] & 128:
        sign = -1
    else:
        sign = 1

Or perhaps just:

    sign = -1 if snapshot[i + 1] & 128 else 1

File: tests/test_basic.py [hunk 13]
```diff
@@ -2,14 +2,153 @@
...
+    def test_get_number_integer(self):
+        for kwargs, exp_result in self.TEST_NUMBERS_INTEGER:
+            self._test_function_integer(exp_result, get_number, **kwargs)
+
+    def test_get_number_float(self):
+        for kwargs, exp_result in self.TEST_NUMBERS_FLOAT:
+            self._test_function_float(exp_result, get_number, **kwargs)
+
+class TextReaderTest(SkoolKitTestCase):
+    TEST_CODES_32_AND_BELOW = (
```
Comment (skoolkid) [inline tests/test_basic.py]: Does TextReaderTest test anything that isn't already tested by BasicListerTest or VariableListerTest? If not, we can remove it. Otherwise consider migrating any useful tests into BasicListerTest or VariableListerTest. (TextReader is more of an implementation detail than a library feature that needs testing directly.)

File: tests/test_basic.py [hunk 14]
```diff
@@ -2,14 +2,153 @@
...
+        ({'snapshot': [121, 45, 171, 159, 85], 'i': 0}, 0.53e-2),        # Positive mantissa, small negative exponent
+        ({'snapshot': [2, 65, 211, 40, 119], 'i': 0}, 0.89e-38),         # Positive mantissa, large negative exponent
+        ({'snapshot': [135, 182, 0, 0, 0], 'i': 0}, -0.91e2),            # Negative mantissa, small positive exponent
+        ({'snapshot': [251, 182, 246, 157, 194], 'i': 0}, -0.76e37),     # Negative mantissa, large positive exponent
+        ({'snapshot': [121, 209, 183, 23, 88], 'i': 0}, -0.64e-2),       # Negative mantissa, small negative exponent
+        ({'snapshot': [2, 182, 239, 144, 158], 'i': 0}, -0.84e-38),      # Negative mantissa, large negative exponent
+        ({'snapshot': [168, 5, 142, 188, 109], 'i': 0}, 0.573625364e12), # Long mantissa
+    )
+
+    def setUp(self):
```
Comment (skoolkid) [inline tests/test_basic.py]: Isn't this method redundant?

File: tests/test_basic.py [hunk 15]
```diff
@@ -2,14 +2,153 @@
...
+        SkoolKitTestCase.setUp(self)
+        self.longMessage = True
+
+    def _test_function_integer(self, exp_result, func, **kwargs):
```
Comment (skoolkid) [inline tests/test_basic.py]: No need for the ``func`` argument here - it's always ``get_number()``.

File: tests/test_basic.py [hunk 16]
```diff
@@ -2,14 +2,153 @@
...
+        kwargs_str = ', '.join(['{}={!r}'.format(k, v) for k, v in kwargs.items()])
+        msg = "{}({}) failed".format(func.__name__, kwargs_str)
+        self.assertEqual(exp_result, func(**kwargs), msg)
+
+    def _test_function_float(self, exp_result, func, **kwargs):
```
Comment (skoolkid) [inline tests/test_basic.py]: See above regarding ``func``.

File: tests/test_basic.py [hunk 17]
```diff
@@ -2,14 +2,153 @@
 import unittest
 
 from skoolkittest import SkoolKitTestCase
-from skoolkit.basic import BasicLister
+from skoolkit.basic import get_number, TextReader, BasicLister, VariableLister
+
+class NumberUtilsTest(SkoolKitTestCase):
```
Comment (skoolkid) [inline tests/test_basic.py]: This is really ``get_number_Test`` now.

File: skoolkit/basic.py (hunk 6 above)
Comment (Wolfe-Lyon) [inline skoolkit/basic.py]: I disagree. I wrote this particular function to mirror the (buggy) way in which the Spectrum handles integer values. While unlikely, it is not impossible that there exists Spectrum software that has been written (intentionally or unintentionally) to rely on this bug to function "correctly". Therefore, outputting integers exactly as the Spectrum sees them is going to be more useful in figuring out how such software works than outputting integers that are technically correct, but not necessarily the same as that software expects.

File: tests/test_basic.py (hunk 13 above)
Comment (Wolfe-Lyon) [inline tests/test_basic.py]: TextReaderTest tests a variety of inputs for the get_chars() method on the TextReader class, which was extracted from the BasicLister class. As a fully	fledged class that is now used by both BasicLister and VariableLister (and may in the future be used elsewhere), TextReader requires its own thorough suite of tests, in my opinion.

Having said that, TextReader's full range of functionality probably is already (indirectly) tested via BasicListerTest and VariableListerTest, so if you disagree with me here that TextReaderTest is necessary, please let me know and I will remove it, or subsume it into BasicListerTest and VariableListerTest as appropriate.

File: skoolkit/basic.py (hunk 7 above)
Comment (Wolfe-Lyon) [inline skoolkit/basic.py]: Fixed in commit 277451935ff7caa5ec1b394de3d140dc1780c7c9.

File: skoolkit/basic.py (hunk 8 above)
Comment (Wolfe-Lyon) [inline skoolkit/basic.py]: Unused variable removed in commit 593dc146cd0c427e87ca7d412231302cb92404b2. Is the expression in the if statement here also unpythonic? I note that there are a few occurrences of integers being explicitly tested for equality with zero throughout SkoolKit so I'm not sure if there's a specific (set of) rule(s) that I should be following.

File: skoolkit/basic.py (hunk 9 above)
Comment (Wolfe-Lyon) [inline skoolkit/basic.py]: Heading removed in commit 37ac97c9281ccf8beeadf8ceaa22f587bdbff2d7.

File: skoolkit/basic.py (hunk 10 above)
Comment (Wolfe-Lyon) [inline skoolkit/basic.py]: Heading removed in commit 37ac97c9281ccf8beeadf8ceaa22f587bdbff2d7.

File: skoolkit/snapinfo.py (hunk 11 above)
Comment (Wolfe-Lyon) [inline skoolkit/snapinfo.py]: Now only called once. Fixed in commit 1a1a6169d854f5d4edfca0e21c29594e84160dec.

File: skoolkit/basic.py (hunk 12 above)
Comment (Wolfe-Lyon) [inline skoolkit/basic.py]: Pythonicity applied and redundant brackets removed in commit 0e626d6530021a525722c4f139968c439404b5a9.

File: tests/test_basic.py (hunk 14 above)
Comment (Wolfe-Lyon) [inline tests/test_basic.py]: Yes. Removed in commit 44c993056eefc6cf836093b6503dfd91508b8dfa.

File: tests/test_basic.py (hunk 15 above)
Comment (Wolfe-Lyon) [inline tests/test_basic.py]: Unnecessary argument removed in commit e0c086ce06cfd1799f1db41c3c5c8e8eb87ae299.

File: tests/test_basic.py (hunk 16 above)
Comment (Wolfe-Lyon) [inline tests/test_basic.py]: Unnecessary argument removed in commit e0c086ce06cfd1799f1db41c3c5c8e8eb87ae299.

File: tests/test_basic.py (hunk 17 above)
Comment (Wolfe-Lyon) [inline tests/test_basic.py]: Name fixed in commit 4fe8757eb38e6e9fc8dd659a40bd1c2a6b98a76d.

**Comment by Wolfe-Lyon** [comment]
Again I've implemented most of your suggested changes and replied to each of your comments above.

Outstanding issues that I have responded to, that you may wish to comment further upon:
- whether or not to get rid of _skip_basic_line()
- behaviour of get_number with integers having a sign byte value which is neither 0 nor 255
- confirmation of your opinion on keeping / removing / breaking up TextReaderTest

**Comment by skoolkid** [comment]
Thanks for making those changes. Regarding the other issues:

* ``_skip_basic_line()`` should be removed and merged into the while loop - see my comment in the code.
* I think assuming that the sign byte of a small integer is valid (i.e. 0 or 255) is good enough. I've never come across a small integer with an invalid sign byte, so I'm reluctant to add support for such a thing.
* TextReader is a private part of BasicLister's and VariableLister's implementation, not part of basic.py's API, so there's no good reason to test it separately. Tests should focus on exercising a library's public API, and make no assumptions about the implementation.

File: skoolkit/basic.py [hunk 18]
```diff
@@ -194,24 +246,112 @@ def _get_num_str(self, j):
...
+                i, line = self._get_char_array_var(i)
+            elif variable_type == 224:
+                # Control variable of a FOR-NEXT loop (111xxxxx)
+                i, line = self._get_control_var(i)
+            elif variable_type == 96:
+                # Number whose name is one letter (011xxxxx)
+                i, line = self._get_short_num_var(i)
+            else:
+                # Basic line (000xxxxx / 001xxxxx)
+                i, line = self._skip_basic_line(i)
```
Comment (skoolkid) [inline skoolkit/basic.py]: This could be replaced with something like:

    else:
        i += get_word(snapshot, i + 2) + 4
        continue

And then there would be no need to check whether ``line`` is blank, I think.

**Comment by Wolfe-Lyon** [comment]
No problem. Thank you for your reply. Issues addressed:

- `_skip_basic_line()` removed (127232463aa58d0165669eaf12bca8361b6feeb6)
- integer handling simplified, assuming sign byte of 0 or 255 (4d1ce6adf4ae44ed919e8071f80c57e68527c00e)
- TextReader's tests removed (32a0217a55deb828a1893b7987fc16cb02d8bc60)

File: skoolkit/basic.py [hunk 19]
```diff
@@ -194,24 +233,105 @@ def _get_num_str(self, j):
...
+        return '\n'.join(lines)
+
+    def _get_string_var(self, i):
+        letter = self.text.get_chars((self.snapshot[i] & 31) + 96)
+        str_length = get_word(self.snapshot, i + 1)
+        value = ''.join([self.text.get_chars(c) for c in self.snapshot[i + 3:i + 3 + str_length]])
+        line = '(String) {}$="{}"'.format(letter, value)
+        return i + 3 + str_length, line
+
+    def _get_num_array_var(self, i):
```
Comment (skoolkid) [inline skoolkit/basic.py]: This method would be clearer if each component of ``line`` were assigned to a suitably named variable, and then ``line`` were built from a single format string. For example:

    ...
    varname = self.text.get_chars(letter)
    ...
    dims = [str(get_word(self.snapshot, c)) for c in range(i, i + 2 * dimensions, 2)]
    ...
    values = [str(get_number(self.snapshot, c)) for c in range(i, i + data_length, 5)]
    ...
    line = '(Number array) {}({})=[{}]'.format(varname, ','.join(dims), ','.join(values))

Use the same approach with ``_get_long_num_var()`` and ``_get_char_array_var()`` too.

File: tests/test_basic.py [hunk 20]
```diff
@@ -2,7 +2,50 @@
 import unittest
 
 from skoolkittest import SkoolKitTestCase
-from skoolkit.basic import BasicLister
+from skoolkit.basic import get_number, TextReader, BasicLister, VariableLister
+
+class get_number_Test(SkoolKitTestCase):
```
Comment (skoolkid) [inline tests/test_basic.py]: I'm thinking this test class should go the same way as TextReaderTest. Is anything being tested here that isn't already tested by BasicListerTest or VariableListerTest? Really, ``get_number()`` is a private function (and so should be renamed ``_get_number()``).

File: skoolkit/basic.py [hunk 21]
```diff
@@ -194,24 +233,105 @@ def _get_num_str(self, j):
...
+                # Array of characters (110xxxxx)
+                i, line = self._get_char_array_var(i)
+            elif variable_type == 224:
+                # Control variable of a FOR-NEXT loop (111xxxxx)
+                i, line = self._get_control_var(i)
+            elif variable_type == 96:
+                # Number whose name is one letter (011xxxxx)
+                i, line = self._get_short_num_var(i)
+            else:
+                # Basic line (000xxxxx / 001xxxxx)
```
Comment (skoolkid) [inline skoolkit/basic.py]: Or:

    #  BASIC line (00xxxxxx)

:)

File: skoolkit/basic.py (hunk 19 above)
Comment (Wolfe-Lyon) [inline skoolkit/basic.py]: Done. I've also updated `_get_string_var()`, `_get_control_var()` and `_get_short_num_var()`
 in the same way to ensure they all look more consistent.

File: tests/test_basic.py (hunk 20 above)
Comment (Wolfe-Lyon) [inline tests/test_basic.py]: Test class dropped. `get_number()` renamed to `_get_number()`.

File: skoolkit/basic.py (hunk 21 above)
Comment (Wolfe-Lyon) [inline skoolkit/basic.py]: Comments are important too!

**Comment by Wolfe-Lyon** [comment]
Latest set of changes made:
- functions clarified (05af3f217f23ba12a22dc00124a1de09ead79c35)
- `get_number()` tests removed. `get_number()` renamed to `_get_number()` (3d7f2f54bef0fb84a773a96e53dbf9ed44d0bbfd)
- comment simplified (05af3f217f23ba12a22dc00124a1de09ead79c35)

**Comment by skoolkid** [comment]
Merged (with minor changes) in f22f9a303dc579bb6862dabf4816c073fbb96332.
